### ISSUE TITLE
Unable to set counter while invoking transactions

### ISSUE THREAD

**Issue by bhanu-ratnakaram** [ecadlabs/taquito#213]
```
Error: 
  message: 'Http error response: (500) [{"kind":"temporary","id":"failure","msg":"Error while applying operation ooJcSdeMGs1wyZye2bTpdSosth2AkEJuwUBU1Z5wvMoUz9QiwSv:\\nbranch refused (Error:\\n                  Counter 43 already used for contract tz1VSUr8wwNhLAzempoch5d6hLRiTh8Cjcjb (expected 44)\\n)"}]\n',
```

This is because when more than one transaction is submitted, the same counter is being used. In other BC platforms like ethereum, we can set this counter while invoking the transaction. Taquito is not allowing us to set counter. IMHO it is quite an important feature to allow setting counter like below

```
const op = await contract.methods.mint(aliceAddress, 100).send({ fee: 30000, gasLimit: 200000, counter: 1})
```

**Comment by jevonearth** [comment]
Thank you @bhanu-ratnakaram we have begun investigating this. We will keep you posted.

**Comment by carte7000** [comment]
@bhanu-ratnakaram we have implemented your suggestion in the https://github.com/ecadlabs/taquito/tree/allow-to-explicitly-set-counter branch

However, we are not sure if it cover your use case, please have a look at this test to confirm that it would solve your problem.
https://github.com/ecadlabs/taquito/compare/allow-to-explicitly-set-counter#diff-e58e08dcf2494864df78d3ffeed46f44R137

I think we might have hit a current limitation of the node and that for now the only solution would be to batch operations together or wait until our previous operation has been included in a block.

Here is an issue related to batching operations #177

**Comment by carte7000** [comment]
FYI this issue might provide complementary information to this one: https://gitlab.com/tezos/tezos/issues/644

**Comment by bhanu-ratnakaram** [comment]
I have tried this branch. I still see the issue persists. Here is the sample ts code to replicate the issue.

import { Tezos } from '@taquito/taquito';
import { InMemorySigner } from '@taquito/signer';
const provider = 'http://localhost:20000';
const signer: any = new InMemorySigner('edsk3QoqBuvdamxouPhin7swCvkQNgq4jP5KZPbwWNnwdZpSpJiEbq');
const f12Contract : any = 'KT1JzhfXHSfvBeudzgERWCemesFnyvVned93'
const aliceAddress: any = 'tz1VSUr8wwNhLAzempoch5d6hLRiTh8Cjcjb'
const bobAddress: any = 'tz1aSkwEot3L2kmUvcoxzjMomb9mvBNuzFK6'
let counter
Tezos.setProvider({ rpc: provider,signer });
​
async function mint() {
    try {
        const contract = await Tezos.contract.at(f12Contract);
        const op = await contract.methods.mint(aliceAddress, 100).send({ fee: 30000, gasLimit: 200000})
        counter = (op.results[0]).counter
        const op2 = await contract.methods.transfer(aliceAddress, bobAddress,100).send({ counter: counter/1 + 1, fee: 30000, gasLimit: 200000 })
        await op.confirmation();
        await op2.confirmation();
    } catch (ex) {
        console.log(ex)
    }
}
​
mint();

**Comment by carte7000** [comment]
Thanks for trying the branch, I am not sure if we can cover this use case until https://gitlab.com/tezos/tezos/issues/36 gets resolved in the node.

In the meantime would awaiting the first operation before emitting the other work for you?
```
import { Tezos } from '@taquito/taquito';
import { InMemorySigner } from '@taquito/signer';
const provider = 'http://localhost:20000';
const signer: any = new InMemorySigner('edsk3QoqBuvdamxouPhin7swCvkQNgq4jP5KZPbwWNnwdZpSpJiEbq');
const f12Contract : any = 'KT1JzhfXHSfvBeudzgERWCemesFnyvVned93'
const aliceAddress: any = 'tz1VSUr8wwNhLAzempoch5d6hLRiTh8Cjcjb'
const bobAddress: any = 'tz1aSkwEot3L2kmUvcoxzjMomb9mvBNuzFK6'
let counter
Tezos.setProvider({ rpc: provider,signer });
​
async function mint() {
try {
const contract = await Tezos.contract.at(f12Contract);
const op = await contract.methods.mint(aliceAddress, 100).send({ fee: 30000, gasLimit: 200000})
counter = (op.results[0]).counter

// Await the first operation confirmation
await op.confirmation();

// Emit the next operation
const op2 = await contract.methods.transfer(aliceAddress, bobAddress,100).send({ counter: counter/1 + 1, fee: 30000, gasLimit: 200000 })
await op2.confirmation();
} catch (ex) {
console.log(ex)
}
}
​
mint();
```

**Comment by bhanu-ratnakaram** [comment]
thats for your response carte.
awaiting is an option but that wont be completely production ready for our use case. This will force us to execute only sequential transaction but never in parallel. 
Regd., issue 36 in tezos, arent the transactions automatically ordered using counter? i.e., if we send 2 transactions with counter values 1 and 2, i thought they would be ordered automatically like ethereum orders using nonce. 
Issue 36 is marked 'nice to have' but if parallel transaction can not be submitted the priority needs escalation i believe.

**Comment by carte7000** [comment]
It is always possible to do a batch operation. You can inject multiple operation with ordered nonce in one operation. However once you inject (one or multiple operation) you must wait for the confirmation before sending another.

Let me know if your use case could be covered by batching your operation together. This is on our roadmap see #177

**Comment by jevonearth** [comment]
@bhanu-ratnakaram Are you still blocked by this issue? Does the new batch operation support in our recent versions address your use-case?

Let me know, and if appropriate, you can close out this issue.

**Comment by bhanu-ratnakaram** [comment]
Hi,

By having a batch operation we can partly resolve the issue i.e., by
building a mechanism to batch the transactions. Hyperledger provides
ordering service to do the same. It would be very useful if this is
provided out of box by tezos
Bhanu Ratnakaram
*Lead Blockchain Developer | Globacap*


322 High Holborn | London WC1V 7PB
<https://www.google.com/maps/place/322+High+Holborn,+Holborn,+London+WC1V+7PB/@51.5180448,-0.1147458,17z/data=!3m1!4b1!4m5!3m4!1s0x48761b4b9bb11c2d:0x94740fea22297042!8m2!3d51.5180448!4d-0.1125571>
(+44) 203 488 0401

**Comment by jevonearth** [comment]
@bhanu-ratnakaram 

Yes, we do have batch support.

For example, you can do something like;

```typescript
Tezos.batch([{kind:'transaction', to: 'tz1a...', amount: 1}, {kind:'transaction', to: 'tz1b...', amount: 1}, {kind:'transaction', to: 'tz1c...', amount: 1}]).send()
```

**Comment by jevonearth** [comment]
Blocked until https://gitlab.com/tezos/tezos/issues/644 is addressed in some way. The Related branch move-require-xhr2-outside-function-scope will remain unmerged until unblocked.

**Comment by pawandhanwani** [comment]
Can we do it like this ?

`const opID = await Tezos.contract.transfer({to : tz1... , amount : 2})
await opID.confirmation()`

**Comment by konchunas** [comment]
> Blocked until https://gitlab.com/tezos/tezos/issues/644 is addressed in some way.
> 
> The Related branch https://github.com/ecadlabs/taquito/tree/allow-to-explicitly-set-counter will remain unmerged until unblocked.

Hey, @jevonearth !
Could you please elaborate how is the mentioned issue blocking this feature?
I'm working on Tezos Gas Station and our backend would greatly benefit from ability to send next transaction while previous one is still being confirmed. Batching for us is not always an option, since transactions might not fit into gas limit.

**Comment by emchristiansen** [comment]
FYI, I'd also benefit from being able to set the counter explicitly. I'm trying to understand mempool semantics by playing with different ways of composing transactions. I'm using the batch API for other stuff but it's not enough for what I'm trying to do now.

**Comment by michaelkernaghan** [comment]
the background issues have been resolved and a similar issue merged.
